File: JAVA/PartitionProblemBackTrackingIterator/src/partitionproblembacktrackingiterator/IteratorCombinations.java
package partitionproblembacktrackingiterator;

public class IteratorCombinations {

    private final int [] combinacion;
    private final int [] num;
    private final int sum;
    private int totalVal;
    private final int valormaximo = 1;
    private final int valorminimo = 0;

    public IteratorCombinations(int[] num, int sum) {
        this.combinacion = new int[num.length];
        this.num = num;
        this.sum = sum;
    }

    public boolean lastCombination() {
        for (int i = 0; i < combinacion.length; i++) {
            if (combinacion[i] == 0) return false;
        }
        return true;
    }

    public boolean nextCombination() {
        for (int i = 0; i < combinacion.length; i++) {
            if (combinacion[i] == valormaximo) {
                combinacion[i] = valorminimo;
            } else {
                combinacion[i] += 1;
                if (isValidCombination()){
                    return true;
                } else {
                    combinacion[i] = valorminimo;
                }
            }
        }
        return false;
    }
    
    public boolean searchedCombination(){
        return (totalVal == sum);
    }

    private boolean isValidCombination() {
        return (totalValues() <= sum);
    }

    private int totalValues() {
        totalVal = 0;
        for (int i = 0; i < num.length; i++) {
            if (combinacion[i] == 1){
                totalVal += num[i];
            }
        }
        return totalVal;
    }
}

File: CPP/PartitionProblemBackTrackingIterator/IteratorCombinations.cpp
#include "IteratorCombinations.h"

IteratorCombinations::IteratorCombinations(int* num, int sum, int size) {
    this->size = size;
    this->combination = new int[size]; 
    this->num = num;
    this->sum = sum;
}

bool IteratorCombinations::lastCombination() {
    for(int i = 0; i < size; i++){
        if(combination[i] == 0) return false;
    }
    return true;
}

bool IteratorCombinations::nextCombination() {
    for(int i = 0; i < size; i++){
        if(combination[i] == maxValue){
            combination[i] = minValue;
        }else{
            combination[i] = maxValue;
            if(isValidCombination()){
                return true;
            }else{
                combination[i] = minValue;
            }
        }
    }
    return false;
}

bool IteratorCombinations::searchedCombination() {
    int actSum = totalValues();
    return (totalVal == sum);
}

bool IteratorCombinations::isValidCombination() {
    return (totalValues() <= sum);
}

int IteratorCombinations::totalValues() {
    this->totalVal = 0;
    for(int i = 0; i < size; i++){
        if(combination[i] == 1){
            this->totalVal += num[i];
        }
    }
    return this->totalVal;
}

